docs/AsyncWrap/README.md

```diff
@@ -1,15 +1,276 @@
+  TCPWRAP: 14,
+  TCPCONNECTWRAP: 15,
+  TIMERWRAP: 16,
+  TLSWRAP: 17,
+  TTYWRAP: 18,
+  UDPWRAP: 19,
+  UDPSENDWRAP: 20,
+  WRITEWRAP: 21,
+  ZLIB: 22 }
+```
+
+##### parent
+
+In some cases the handle was created from another handle object. In those
+cases the `parent` argument is set the creating handle object. If there is
+no parent handle then it is just `null`.
+
+The most common case is the TCP server. The TCP server itself is a `TCP` handle,
+but when receiving new connection it creates another `TCP` handle that is
+responsible for the new socket. It does this before emitting the `onconnection`
+handle event, thus the asyncWrap hooks are called in the following order:
+
+```
+```javascript
+init // TCP (socket)
+before // TCP (server)
+after // TCP (server)
+```
+
+This means it is not possible to know in what handle context the new socket
+handle was created using the `before` and `after` hooks. However the
+`parent` argument provides this information.
+
+## example
+
+A classic use case for AsyncWrap is to create a long-stack-trace tool.
+
+```javascript
+const asyncWrap = process.binding('async_wrap');
+
+asyncWrap.setupHooks(init, before, after);
+asyncWrap.enable();
+
+// global state variable, that contains the current stack trace
+let currentStack = '';
+
+function init(provider, parent) {
+  // When a handle is created, collect the stack trace such that we later
+  // can see what involved the handle constructor.
+  const localStack = (new Error()).stack.split('\n').slice(1).join('\n');
+
+  // Compute the full stack and store it as a property on the handle object,
+  // such that it can be fetched later.
+  const extraStack = parent ? parent._full_init_stack : currentStack;
+  this._full_init_stack = localStack + '\n' + extraStack;
+}
+function before() {
+  // A callback is about to be called, update the `currentStack` such that
+  // it is correct for when another handle is initialized or `getStack` is called.
+  currentStack = this._full_init_stack;
+}
+function after() {
+  // At the time of writing there are some odd cases where there is no handle
+  // context, this line prevents that from resulting in wrong stack trace. But
+  // the stack trace will be shorter compared to what ideally should happen.
+  currentStack = '';
+}
+
+function getStack(message) {
+  const localStack = new Error(message);
+  return localStack.stack + '\n' + currentStack;
+}
+module.exports = getStack;
+```
+
+Please note that this example is way simpler, than what is required from a
+complete long-stack-trace implementation.
+
+## Things you might not expect
+
+* It is not obvious when a handle object is created. For example the TCP server
+creates the `TCP` handle when `.listen` is called and it may perform an DNS
+lookup before that.
+
+* `console.log` is async and thus invokes AsyncWrap, thus using `console.log`
+inside one of the hooks, creates an infinite recursion. Use `fs.syncWrite(1, msg)`
+or `process._rawDebug(msg)` instead.
```

Comment:
possibly mention that `fs.syncWrite()` doesn't call `util.inspect()` so the output will be unformatted? Not a big thing.